File: 2ºAÑO/1ºCuatrimestre/PDOO/examen_p1_p2/examen/lib/jugador.rb
# encoding:utf-8

module Civitas
  class Jugador
    
      @@saldo_inicial = 7500
      @@paso_por_salida = 1000
      @@precio_libertad = 200
      @@casas_max = 4
      @@casas_por_hotel = 4
      @@hoteles_max = 4

    
     def initialize(nombre)
      @encarcelado = false
      @nombre = nombre
      @num_casilla_actual = 0
      @puede_comprar = false
      @saldo = 0
      @propiedades = Array.new
      @salvoconducto = nil
    end
    
     #creo que está mal
    def new_copia(otro)
      @nombre = otro.nombre
      @encarcelado = otro.is_encarcelado
      @num_casilla_actual = otro.num_casilla_actual
      @puede_comprar = otro.puede_comprar
      @saldo = otro.saldo
      @propiedades = otro.propiedades
      @salvoconducto = otro.salvoconducto
    end
    
    def <=>(otro) # #1 if self>other; 0 if self==other; -1 if self<other
      @saldo <=> otro.saldo
    end
    
=begin   
    def cancelar_hipoteca(ip)
      
    end
=end   
    def cantidad_casas_hoteles
      aux = 0
      
      @propiedades.each do |i|
        aux += i.cantidad_casas_hoteles
      end
      
      return aux
    end
=begin    
    def comprar(titulo)
      
    end
    
    def construir_casa(ip)
      
    end
    
    def construir_hotel(ip)
    
    end
=end
    
    def debe_ser_encarcelado
      if is_encarcelado
        return false
      else
        if !tiene_salvoconducto
          return true
        else
          perder_salvoconducto
          Diario.instance.ocurre_evento("El jugador se libra de la cárcel")
          return false
        end
      end
    end
    
    def en_bancarrota
      return @saldo < 0      
    end
    
    def encarcelar(num_casilla_carcel)
      if debe_ser_encarcelado
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("Jugador encarcelado")
      end
      return @encarcelado
    end
    
    def existe_la_propiedad(ip)
      return @propiedades.at(ip)
    end
=begin   
    def hipotecar(ip)
      
    end
=end   
    def is_encarcelado
      return @encarcelado
    end
    
    def modificar_saldo(cantidad)
      @saldo +=cantidad
      true
    end
    
    def mover_a_casilla(num_casilla)
      if is_encarcelado
        return false
      else
        @num_casilla_actual = num_casilla
        @puede_comprar = false
        Diario.instance.ocurre_evento("Jugador movido de casilla")
        return true
      end
    end
    
    def obtener_salvoconducto(sorpresa)
      if is_encarcelado
        return false
      else
        @salvoconducto = sorpresa
        return true
      end
    end
    
    def paga(cantidad)
      return modificar_saldo(cantidad*(-1))
    end
    
    def paga_alquiler(cantidad)
      if is_encarcelado
        return false
      else
        return paga(cantidad)
      end
    end
    
    def paga_impuesto(cantidad)
      if is_encarcelado
        return false
      else
        return paga(cantidad)
      end
    end
    
    def pasa_por_salida
      modificar_saldo(@@paso_por_salida)
      Diario.instance.ocurre_evento("El jugador pasa por la salida y cobra")
      return true
    end
    
    def perder_salvoconducto
      @salvoconducto.usada
      @salvoconducto = nil
    end
    
    def puede_comprar_casilla
      if is_encarcelado
        @puede_comprar = false
      else
        @puede_comprar = true
      end
      return @puede_comprar
    end
    
    def puede_salir_carcel_pagando
      return @saldo >= @@precio_libertad
    end
    
    
=begin
def puede_edificar_casa(titulo)
        return titulo.construir_casa(self)  #¿self?
      end
    
  
      def puede_edificar_hotel(titulo)
        return titulo.construir_hotel(self)
      end
=end
    
    
    
    def puedo_gastar(precio)
      if is_encarcelado
        return false
      else
        return @saldo >= precio
      end
    end
    
    def recibe(cantidad)
      if is_encarcelado
        return false
      else
        return modificar_saldo(cantidad)
      end
    end
    
    def salir_carcel_pagando
      if is_encarcelado && puede_salir_carcel_pagando
        paga(@@precio_libertad)
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador paga y sale de la cárcel")
        return true
      else
        return false
      end
    end
    
    def salir_carcel_tirando
      if Dado.instance.salgo_de_la_carcel
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador sale de la cárcel")
        return true
      else
        return false
      end
    end
    
    def tiene_algo_que_gestionar
      return @propiedades.any? #duda
    end
    
    def tiene_salvoconducto
      return @salvoconducto != nil
    end
    
    def to_s
      "Jugador:\n" + "\nencarcelado =" + @encarcelado.to_s + "\nnombre =" << @nombre.to_s + "\nnum_casilla actual =" + @num_casilla_actual.to_s + 
        "\npuede_comprar =" + @puede_comprar.to_s + "\nsaldo =" + @saldo.to_s + "\npropiedades:\n" + @propiedades.to_s
    end
    
    def vender(ip)
      if is_encarcelado
        return false
      else
        @titulo = nil
        if existe_la_propiedad(ip) && @titulo.vender(self)
          @propiedades.remove(ip)
          Diario.instance.ocurre_evento("Propiedad vendida con éxito")
          return true
        else
          return false
        end
      end
    end
    
     attr_reader :casas_max, :casas_por_hotel, :hoteles_max, 
      :nombre, :num_casilla_actual, :precio_libertad,
      :paso_por_salida, :propiedades, :puede_comprar, :saldo, :salvoconducto
    
    protected :debe_ser_encarcelado, :new_copia
    private :existe_la_propiedad, :perder_salvoconducto,
      :puede_salir_carcel_pagando, :puedo_gastar
    
  end
end

File: 2ºAÑO/1ºCuatrimestre/PDOO/Civitas/Civitas/src/civitas/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

import java.lang.Float;

/**
 *
 * @author Enrique
 */
public class Jugador implements Comparable<Jugador> {
    static protected int CasasMax = 4;
    static protected int CasasPorHotel = 4;
    protected boolean encarcelado = false;
    static protected int HotelesMax = 4;
    private String nombre;
    private int numCasillaActual;
    static protected float PasoPorSalida = 1000;
    static protected float PrecioLibertad = 200;
    private boolean puedeComprar;
    private float saldo;
    static private float SaldoInicial = 7500;
    private ArrayList<TituloPropiedad> propiedades = new ArrayList<>();
    Sorpresa salvoconducto;
    
    void setSaldo(int i){
        saldo = i;
    }
    
    Jugador(String nombre) {
        this.nombre = nombre;
    }
    
    protected Jugador(Jugador otro){
        encarcelado = otro.encarcelado;
        nombre = otro.nombre;
        numCasillaActual = otro.numCasillaActual;
        puedeComprar = otro.puedeComprar;
        saldo = otro.saldo;
        propiedades = otro.propiedades;
        salvoconducto = otro.salvoconducto;
    }
    
    /*boolean cancelarHipoteca(int ip) {
        
    }*/
    
    int cantidadCasasHoteles() {
        int aux=0;
        for(int i = 0; i< propiedades.size(); i++)
        {
            aux += aux + propiedades.get(i).cantidadCasasHoteles();
        }
        return aux;
    }
    
    
    
    /*equal to 0: Object f is equal to the argument object
    less than 0: Object f is less than the argument object
    greater than 0: Object f is greater than the argument object*/

    
    @Override
    public int compareTo(Jugador arg0) {
        Float aux = this.getSaldo();
        return aux.compareTo(arg0.getSaldo());
    }
    
    /*boolean comprar(TituloPropiedad titulo) {
        
    }*/
    
    
    protected boolean debeSerEncarcelado() {
        if(this.isEncarcelado()) {
            return false;
        }
        
        else {
            if(!tieneSalvoConducto()) {
                return true;
            }
            
            else {
                this.perderSalvoConducto();
                Diario.getInstance().ocurreEvento("El jugador se libra de la cárcel ");
                return false;
            }
        }
    }
    
    boolean enBancarrota() {
        return saldo < 0;
    }
    
    boolean encarcelar(int numCasillaCarcel) {
        if(this.debeSerEncarcelado()) {
            this.moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.getInstance().ocurreEvento("Jugador encarcelado");
        }
        return encarcelado;
    }
    
    private boolean existeLaPropiedad(int ip) {
        return propiedades.contains(ip);
    }
    
    private int getCasasMax() {
        return CasasMax;
    }
    
    int getCasasPorHotel() {
        return CasasPorHotel;
    }
    
    private int getHotelesMax() {
        return HotelesMax;
    }
    
    protected String getNombre() {
        return nombre;
    }
    
    int getNumCasillaActual() {
        return numCasillaActual;
    }
    
    private float getPrecioLibertad() {
        return PrecioLibertad;
    }
    
    private float getPremioPorSalida() {
        return PasoPorSalida;
    }
    
    protected ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }
    
    boolean getPuedeComprar() {
        return puedeComprar;
    }
    
    protected float getSaldo() {
        return saldo;
    }
    
    /*boolean hipotecar(int ip) {
        
    }*/
    
    public boolean isEncarcelado() {
        return encarcelado;
    }
    
    boolean modificarSaldo(float cantidad){
        saldo +=cantidad;
        return true;
    }
    
    boolean moverACasilla(int numCasilla){
        if(this.isEncarcelado()) {
            return false;
        }
        
        else {
            numCasillaActual = numCasilla;
            puedeComprar = false;
            Diario.getInstance().ocurreEvento("Jugador movido de casilla");
            return true;
        }
    }
    
    boolean obtenerSalvoconducto(Sorpresa sorpresa){
        if(this.isEncarcelado()) {
            return false;
        }
        
        else {
            salvoconducto = sorpresa;
            return true;
        }
    }
    
    boolean paga(float cantidad){
        return this.modificarSaldo(cantidad*(-1));
    }
    
    boolean pagaAlquiler(float cantidad){
        if(this.isEncarcelado()) {
            return false;
        }
        
        else {
            return this.paga(cantidad);
        }
    }
    
    boolean pagaImpuesto(float cantidad){
        if(this.isEncarcelado()) {
            return false;
        }
        
        else {
            return this.paga(cantidad);
        }
    }
    
    boolean pasaPorSalida(){
        this.modificarSaldo(this.getPremioPorSalida());
        Diario.getInstance().ocurreEvento("El jugador pasa por la salida y cobra");
        return true;
    }
    
    void perderSalvoConducto(){
        salvoconducto.usada();
        salvoconducto = null;
    }
    
    boolean puedeComprarCasilla(){
        if(this.isEncarcelado()) {
            puedeComprar = false;
        }
        
        else {
            puedeComprar = true;
        }
        
        return puedeComprar;
    }
    
    boolean puedeSalirCarcelPagando(){
        return saldo >= PrecioLibertad;
    }
    
    /*boolean contruirCasa(TituloPropiedad titulo){
        return titulo.contruirCasa(this);
    }
    
    boolean construirHotel(TituloPropiedad titulo){
        return titulo.construirHotel(this);
    }*/
    
    boolean puedoGastar(float precio){
        if(this.isEncarcelado()) {
            return false;
        }
        
        else {
            return saldo >= precio;
        }
    }
    
    boolean recibe(float cantidad){
        if(this.isEncarcelado()) {
            return false;
        }
        
        else {
            return this.modificarSaldo(cantidad);
        }
    }
    
    boolean salirCarcelPagando(){
        if(this.isEncarcelado() && this.puedeSalirCarcelPagando()) {
            this.paga(PrecioLibertad);
            encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador paga y sale de la cárcel");
            return true;
        }
        
        else {
            return false;
        }
    }
    
    boolean salirCarcelTirando(){
        if(Dado.getInstance().salgoDeLaCarcel()) {
            encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador sale de la cárcel");
            return true;
        }
        
        else {
            return false;
        }
    }
    
    boolean tieneAlgoQueGestionar(){
        return !this.getPropiedades().isEmpty();
    }
    
    boolean tieneSalvoConducto(){
        return salvoconducto != null;
    }

    @Override
    public String toString() {
        return "Jugador{" + "encarcelado=" + encarcelado + ", nombre=" + nombre + ", numCasillaActual=" + numCasillaActual + 
                ", puedeComprar=" + puedeComprar + ", saldo=" + saldo + ", propiedades=" + propiedades + '}';
    }
    
    boolean vender(int ip){
        if(this.isEncarcelado()) {
            return false;
        }
        
        else {
            TituloPropiedad titulo = null;
            if(this.existeLaPropiedad(ip) && titulo.vender(this)) {
                propiedades.remove(ip);
                Diario.getInstance().ocurreEvento("Propiedad vendida con éxito");
                return true;
            }
            
            else {
                return false;
            }
        }
    }
}